### PR TITLE
Add negotiator argument & reason to close

### PR DESCRIPTION
**Pull request by ghost**
This PR adds the possibility to hook into the Websocket Handshake by providing a negotiator themself through the constructor, it also adds the reason to the close event and removes the close exception (close codes >= 2000 are by definition NOT errors).

### CODE REVIEW

File: src/WebSocketMiddleware.php [hunk 1]
```diff
@@ -12,14 +12,23 @@
 final class WebSocketMiddleware
 {
     private $paths;
-    private $connectionHandler = null;
+    private $connectionHandler;
     private $subProtocols;
+    private $negotiator;
 
-    public function __construct(array $paths = [], callable $connectionHandler = null, array $subProtocols = [])
+    public function __construct(array $paths = [], callable $connectionHandler = null, array $subProtocols = [], $negotiator = null)
```
Comment (mbonneau) [inline src/WebSocketMiddleware.php]: This should be typed `NegotiatorInterface`

Comment (ghost) [reply src/WebSocketMiddleware.php]: @mbonneau Added.

**Comment by ghost** [comment]
@mbonneau ping

**Comment by mbonneau** [comment]
@CharlotteDunois I agree with the close code update, but what is the reason that the Negotiator needs to be injected?

**Comment by ghost** [comment]
@mbonneau I need to be able to read the HTTP headers and don't initiate a websocket upgrade at all when the HTTP request is no good (Authentication in my case).

**Comment by mbonneau** [comment]
@CharlotteDunois Because this is middleware, would it be more appropriate to take care of authentication before the `WebSocketMiddleware` is even invoked?

**Comment by ghost** [comment]
@mbonneau That might work for my use case, but there are other uses where being able to inject a negotiator would improve the usability of this middleware. Others might have a very optimized handshake process for their application and being able to inject a custom negotiator will allow them to use this middleware as base for a websocket server, instead of having to re-implement the same thing over again. 

In the end I don't see much damage being done in being able to inject a negotiator, even though it might be unused for most use cases.

**Pull request closed without merging**